tests/topotests/pbr_topo1/test_pbr_topo1.py

```diff
@@ -399,6 +399,116 @@ def _get_router_rules(router, expected):
+            return "Failed to parse nexthop-groups JSON"
+
+        logger.info("Got nexthop-groups JSON: %s" % nhg_data)
+
+        if not nhg_data:
+            return "Nexthop-group C data is empty"
+
+        # Find the nexthop-group with name "C" (ID is variable)
+        nhg = None
+
+        if isinstance(nhg_data, list):
+            for item in nhg_data:
+                if isinstance(item, dict) and item.get("name") == "C":
+                    nhg = item
+                    break
+        elif isinstance(nhg_data, dict):
+            for key, value in nhg_data.items():
+                if isinstance(value, dict) and value.get("name") == "C":
+                    nhg = value
+                    break
+
+        if nhg is None:
+            return "Nexthop-group C not found in output"
+
+        if nhg.get("valid") != valid:
+            return "Nexthop-group C valid=%s, expected=%s" % (nhg.get("valid"), valid)
+
+        if not valid:
+            # If nexthop-group is invalid, we don't need to check nexthops
+            return None
+
+        nexthops = nhg.get("nexthops", [])
+        if len(nexthops) == 0:
+            return "Nexthop-group C has no nexthops"
+
+        for nexthop in nexthops:
+            if nexthop.get("targetVrf") != "vrf-chiyoda":
+                return "Nexthop %s is not in vrf-chiyoda" % nexthop.get("nexthop")
+            if nexthop.get("valid") != valid:
+                return "Nexthop %s valid=%s, expected=%s" % (nexthop.get("nexthop"), nexthop.get("valid"), valid)
+
+        return None
+
+    logger.info("Step 1: Verifying nexthop-group C is valid")
+    test_func = partial(check_nhg_c_valid, router, True)
+    _, result = topotest.run_and_expect(test_func, None, count=10, wait=1)
+    assertmsg = "Failed to verify nexthop-group C is valid initially"
+    if result is not None:
+        logger.error(result)
+    assert result is None, assertmsg
+
+    logger.info("Step 2: Disabling VRF vrf-chiyoda")
+    router.run("ip link set dev r1-eth3 nomaster")
+    router.run("ip link set vrf-chiyoda down")
+    router.run("ip link del dev vrf-chiyoda")
+
+    # Allow time for VRF deletion to propagate
+    topotest.sleep(2, "Waiting for VRF deletion to propagate")
+
+    logger.info("Step 3: Verifying nexthop-group C is invalid after VRF deletion")
+    test_func = partial(check_nhg_c_valid, router, False)
+    _, result = topotest.run_and_expect(test_func, None, count=10, wait=1)
+    assertmsg = "Failed to verify nexthop-group C is invalid after VRF deletion"
+    if result is not None:
+        logger.error(result)
+    assert result is None, assertmsg
+
+    logger.info("Step 4: Re-enabling VRF vrf-chiyoda")
+    router.run("ip link add vrf-chiyoda type vrf table 1000")
+    router.run("ip link set dev r1-eth3 master vrf-chiyoda")
+    router.run("ip link set vrf-chiyoda up")
+
+    # Allow time for VRF restoration to propagate
+    topotest.sleep(2, "Waiting for VRF restoration to propagate")
```

Comment:
If we have the run and expect below, what is a sleep buying us here?  We do not want random sleeps in the tests they need to be part of the code that is looking for the state to be achieved